### ISSUE TITLE
Document contains wrong value for key

### ISSUE THREAD

**Issue by blaugold** [cbl-dart/cbl-dart#329]
Hi @blaugold,

we have a probably similar problem with cbl 1.0.3. Sometimes the value (List type) of a specific key is copied to another key in every document. However, this problem occurs only at around ~3/4 of all initial database syncs. It is very hard to determine external factors, that influence the wrong behavior.

This is the usual document structure for all database items
![Bildschirmfoto 2022-03-17 um 02 23 36](https://user-images.githubusercontent.com/82163574/158718962-c52bb0c5-a68b-4358-a205-25da76545461.png)

However, at 50% of all initial synchronization, the value of the map entry 9 is copied to entry 0 (the behavior applied to all 60 replicated documents).
![Bildschirmfoto 2022-03-17 um 02 29 50](https://user-images.githubusercontent.com/82163574/158719237-cb1624b8-8c9a-41e8-9048-0e5ab54ecf86.png)

The documents are directly loaded from the database after being replicated. The `await db.document(documentId)` function is called by the only existing change listener of the database. However, we have a second cbl instance with a separate change listener running in parallel but without a replicator. The only thing we observe (but not sure if it's relevant) is that usually the problem only occurs if both databases are initialized at app startup. If we only delete and re-sync the replicated database and leave the other database untouched, then the mentioned behavior is much more unlikely. I don't see any strange behavior in the database without replicator.
I hope, the error description can give you some idea of the problem. If you need any further information, pls let me know.

_Originally posted by @nblum37 in https://github.com/cbl-dart/cbl-dart/issues/322#issuecomment-1069823552_

**Comment by blaugold** [comment]
@nblum37
Can you tell whether this issue appeared with a specific version of `cbl`?

Are you using `Document.toPlainMap` or the getters (e.g `doc.string('foo')`) to access the document content?
I'd like to know if using one instead of the other makes any difference. For diagnosing this, it's important that you only use one mechanism with the same document, since the document is doing some caching internally when loading data.

**Comment by blaugold** [comment]
I've just publish `1.1.0` of `cbl_dart`, `cbl_flutter_ce` and `cbl_flutter_ee`. They contain a fix for an issue which could be related.

**Comment by nblum37** [comment]
I'm currently checking different versions of cbl. Is there a specific version I should check? As far as I can remember, I never observed this issue in the 1.0.0-betaX versions. 
I access the first value with a string getter first (`doc.string('foo')`). The summary was plotted afterward by using the following code
```
        Map<String, Object?> docMap = doc.toPlainMap();
        print('Document entry types:');
        int index = 0;
        for (MapEntry entry in docMap.entries) {
          print('$index = {map entry} <${entry.key.runtimeType}> -> <${entry.value.runtimeType}>');
          index++;
        }
```
Some interesting findings:
1. If I access the document with `doc.toPlainMap()` first, the getters afterwards receive the right value.
2. If I access the getter before the `doc.toPlainMap()`, I sometimes run into the issue with the wrong values

**Comment by nblum37** [comment]
I will check out the new version directly

**Comment by blaugold** [comment]
> I only see cbl 1.0.3 available at pub.dev

The new versions are for the platform-specific packages:
 - cbl_dart
 - cbl_flutter_ce
 - cbl_flutter_ee

> Some interesting findings:
> 
> 1. If I access the document with `doc.toPlainMap()` first, the getters afterwards receive the right value.
> 2. If I access the getter before the `doc.toPlainMap()`, I sometimes run into the issue with the wrong values

That is very helpful. The getters use an optimization which creates an object that is used to speed up dictionary lookups. If a key is used often enough, this object is created and used for subsequent lookups. I'll have to look into what's going on there.

**Comment by blaugold** [comment]
Are you using queries at some point with the database that is being replicated?
If you are, #331 is likely to fix this issue.
You can try it out by using this in your `pubspec.yaml`:
```yaml
  cbl:
    git:
      url: https://github.com/cbl-dart/cbl-dart
      ref: fix/query-dict-keys
      path: packages/cbl
```

**Comment by nblum37** [comment]
I updated to
```
  cbl: ^1.0.3  # License: Apache 2.0
  cbl_flutter: ^1.0.0+2  # License: Apache 2.0
  cbl_flutter_ce: ^1.1.0  # License: Apache 2.0
```
and performed a flutter clean before running the next build. I still see the same situation.

What cbl_flutter and cbl_flutter_ce version should I use for
```
  cbl:
    git:
      url: https://github.com/cbl-dart/cbl-dart
      ref: fix/query-dict-keys
      path: packages/cbl
```
?

**Comment by blaugold** [comment]
> What cbl_flutter and cbl_flutter_ce version should I use for

There's no need to make any changes to those.

**Comment by nblum37** [comment]
Unfortunately, pub get fails.

```
  cbl:
    git:
      url: https://github.com/cbl-dart/cbl-dart
      ref: fix/query-dict-keys
      path: packages/cbl
  cbl_flutter: ^1.0.0+2  # License: Apache 2.0
  cbl_flutter_ce: ^1.1.0  # License: Apache 2.0
```

![Bildschirmfoto 2022-03-17 um 13 34 01](https://user-images.githubusercontent.com/82163574/158809546-97934258-f8cf-4f75-88bc-18932826d55e.png)

**Comment by blaugold** [comment]
Sorry about that. This should work:
```yaml
dependency_overrides:
  cbl:
    git:
      url: https://github.com/cbl-dart/cbl-dart
      ref: fix/query-dict-keys
      path: packages/cbl
```

**Comment by nblum37** [comment]
Ah ok, I also realized, that the order of the key at toPlainMap() depends on what key it did access before with getters.
Also if I access to keys via getters before the toPlainMap function the issue seems to be key dependent and not getter order dependent. So
```
String? a = doc.string('a');
String? b = doc.string('b');
```
or 
```
String? b = doc.string('b');
String? a = doc.string('a');
```
only affects the result for key `a`.

The new fix does not seem to solve the issue.

**Comment by nblum37** [comment]
Yes, I'm also already thinking about how to provide you some more information to reproduce the issue or what kind of minimal example can help. Probably I'll need a little bit more time to dig deeper into the exact environment, which increases the error probability.

**Comment by nblum37** [comment]
Ok, what I can say by now:
It seems that something goes already wrong during database initialization. If I close and re-open the database (without deleting the replicated database and not touching the second database), the problem disappeared.

Besides of that, I get the following behavior:
When calling
```
String? a = doc.string('a');
Map<String, dynamic> docMap = doc.toPlainMap();
```
I get a wrong result for the value of key `a` in both statements.

When calling
```
Map<String, dynamic> docMap = doc.toPlainMap();
String? a = doc.string('a');
```
I get a correct result for the value of key `a` in both statements.

I can repeat both approaches as much as I want, they still behave the same way. 
However, if I simply close and re-open the database, both calls produce the right results. In addition, restarting the App does not reproduce the wrong behavior. I have a strong feeling, that the issue is related to the first time the database is created and opened.

**Comment by nblum37** [comment]
Some more insights:
The issue remains also if other documents are replicated later after the initial replication. However, directly re-opening the database after the first creation (as a naive workaround) did not solve the issue. So re-opening the database only solves the issue at a later point in time.
As written above, recreating the database only causes the issue if the second database is created as well. This seems to be somehow a requirement for the issue to happen.

So in order to reproduce the issue with a ~50% probability, I need to
1. Deleted both databases
2. Create both databases and activate replicator for one of them (in sequential or parallel order)

The problem persists also for newly replicated documents. To resolve the problem including newly replicated documents, I need to
1. Close the replicated database (but not directly after creation; it is unclear how much time to wait)
2. Open the replicated database (and start its replicator)

This is really strange...
So the only "workaround" right now is to start replicating and after 3s I automatically re-open the DB again :D I'll continue investigating.

**Comment by blaugold** [comment]
Thanks for digging into this! Especially, the interaction between the two database instances is strange.

I've created a branch which adds some asserts and logging to debug how dictionary entries are looked up:
```yaml
dependency_overrides:
  cbl:
    git:
      url: https://github.com/cbl-dart/cbl-dart
      ref: debug/329/shared-key-collision
      path: packages/cbl
```
Could you trigger the issue and report back what is being logged?

**Comment by nblum37** [comment]
Thank you for the debug package! However, I don't get any additional logs when I run 
```
Document doc = await db.document(documentId);
String? a = doc.string('a');
Map<String, dynamic> docMap = doc.toPlainMap();
```

Basically, I don't get any additional logs except my own log statements regarding the command execution. I used 
```
  Database.log
    ..console.level = LogLevel.debug
    ..custom = DartConsoleLogger(LogLevel.debug);
```
for the database initialization. I get a lot of CBL debug messages for all the other database-related operations like queries, replicator, ...

**Comment by nblum37** [comment]
Ah wait, probably the values were already cached before

**Comment by nblum37** [comment]
Hmm ok, even at the first  `Document doc = await db.document(documentId);` there is no additional output.
This is the result for the two database access methods:
![Bildschirmfoto 2022-03-18 um 17 43 03](https://user-images.githubusercontent.com/82163574/159046052-2db8caf0-20c0-4a5e-ad17-e3f2772b0c66.png)
'document_type' is the key, that I used for testing purposes and mentioned as 'a' before.

**Comment by blaugold** [comment]
I'm just using `print` for these debug log messages, so there's no need to change anything related to `Database.log`.
You should get a log message like this one, the third time the same key is accessed through a getter:
```
OptimizedDictKey(0x174d60fc): Added optimized key for "document_type"
```
Every instance of `Database` has its own set of `OptimizedDictKey`s, which are populated when getters or setters are used. If you don't get any of those log messages, I suspect that the dependency override is not getting picked up. Maybe running again `flutter pub get` helps.

If 60 documents are replicated and accessed through getters, we should definitely see the `OptimizedDictKey`s being added. Let's verify that you get the log messages with something like this:
```dart
final doc = MutableDocument();
await db.saveDocument(doc);
final loadedDoc = await db.document(doc.id);
loadedDoc!.string("foo");
loadedDoc.string("foo");
loadedDoc.string("foo");
```

**Comment by nblum37** [comment]
When I sync the whole application, I find the following two lines in the log
![Bildschirmfoto 2022-03-18 um 18 54 13](https://user-images.githubusercontent.com/82163574/159057293-1ba6b0a2-e560-4e14-8064-0a73c3d25022.png)

**Comment by nblum37** [comment]
The interesting thing is, that "Added optimized key for " is called eight times. However, the second line "Now using shared key " can only be found for two keys, "document_type" and "asset_type".

**Comment by nblum37** [comment]
Another interesting finding. After I re-open the database (and everything is working), I find only one "Added optimized key for" entry and it is for the key `document_type` with a different shared key
![Bildschirmfoto 2022-03-18 um 19 13 57](https://user-images.githubusercontent.com/82163574/159060320-1e3cd02f-8a35-467b-ba38-d0f7e60c15c5.png)
.

**Comment by blaugold** [comment]
That the shared key for the same key changes confirms what I was suspecting regarding how a wrong value from the same document is returned for a key.

Shared keys are small integers that are used instead of the full key string when persisting documents. Every database has a table that maps the shared keys to the full key string. `OptimizedDictKey`s find and store the shared key for their full key the first time they are used. This makes subsequent lookups a lot faster. We create and cache `OptimizedDictKey`s for keys that are used multiple times.  Somehow, these `OptimizedDictKey`s are getting poisoned with the wrong shared key in certain conditions. When a database is reopened, a new empty `OptimizedDictKey`s cache is used, in which the `OptimizedDictKey`s then seems to work correctly.

When documents are inserted into the database the shared keys table is updated. If a transaction is rolled back so is the shared keys table. Maybe the first document access happens during a transaction that is then rolled back.

**Comment by nblum37** [comment]
Ok perfect. I'm still trying to get more information about how this strange dependency from the second database influences the behavior. One additional add-on:
Assume that both DBs are working fine after app start-up. When I start re-creating the replicated database, the issue occurrence probability is at >50%. If I re-create the replicated database while the other DB is closed, the issue rate seems to be zero. Unfortunately, everything is really stochastic... So it's really hard, to provide real evidence.

**Comment by blaugold** [comment]
I don't think there is necessarily any direct interaction between the two databases. Just the time it takes to open the other database could change the timing of code execution slightly and open a window for the poisoning of the `OptimizedDictKey`.

**Comment by nblum37** [comment]
Yes, I fully agree. I also thought about some jitter in the timing, which could be a reasonable noise generator for the error pattern.

**Comment by blaugold** [comment]
I have opened an [issue](https://github.com/couchbase/couchbase-lite-C/issues/287) in CBL C for what, I think, is the root cause.

One other thing that would be interesting is if you can find any CBL log messages like:
```
20:20:22.348567| [DB] verbose: {DB#1} abort transaction
```
when the issue is occurs.

BTW: On android, the `Database.log.console` logger is not hooked up to anything. `Database.log.custom!.level = LogLevel.debug` should be fine.

**Comment by nblum37** [comment]
I'm really sry, but I'm not able to find any "abort transaction"

**Comment by nblum37** [comment]
Btw: For some reason, the wrongly shared key is sometimes not "64" but "65" in very rare cases. But I guess this is also related to some timing jitter.

**Comment by blaugold** [comment]
OK, that's interesting.

I created another branch which disables the use of `OptimizedDictKey`s:
```yaml
dependency_overrides:
  cbl:
    git:
      url: https://github.com/cbl-dart/cbl-dart
      ref: debug/329/disable-opt-dict-keys
      path: packages/cbl
```
If the issue still persists, the problem is somewhere else, and it just manifests in bad `OptimizedDictKey`s.

Is the other database that is not being replicated new and empty and if not, did it at any point contain documents with the `document_type` key?

Something else that would be helpful are the Couchbase Lite logs from a run where the issue was visible.
The best way to capture those is file logging:
```dart
import 'package:path_provider/path_provider.dart';

Database.log.file
  ..config = LogFileConfiguration(
    directory: (await getApplicationSupportDirectory()).path,
    usePlainText: true,
  )
  ..level = LogLevel.verbose;
```
You can use the Android Studio Device File Explorer to get the logs from `/data/data/<package_name>`.
It doesn't have to be publicly if the logs are sensitive. My email is gabriel@terwesten.net.

**Comment by nblum37** [comment]
For some reason, Pub get can't find the reference "debug/329/disable-opt-dict-keys" at the moment? Same for me online at GitHub.

**Comment by blaugold** [comment]
Sorry, forgot to push the branch. It should be available now.

**Comment by nblum37** [comment]
Ok perfect. So the error does not occur with the optimized dict being deactivated :)

The second database contains summaries of the replicated database also including the `document_type` key. With every new document in the replicated database, the summary in the other database is refreshed. As side information: The value for `document_type` in the replicated database is always "asset_chain" and in the summary database it is always "asset_main". From a timing perspective, document_type will always first appear in the replicated database. However, if the issue occurs, the summary documents are never generated, as the `document_type` values are all invalid.

Yes, I can provide you with some logs. Do you want to update the `debug/329/shared-key-collision` branch with some additional debug output according to the new findings first?

**Comment by blaugold** [comment]
Thanks for check that. 

> Yes, I can provide you with some logs. Do you want to update the debug/329/shared-key-collision branch with some additional debug output according to the new findings first?

Good idea. I have updated the `debug/329/shared-key-collision` branch to log directly to the CBL logs and added a few log statements.

**Comment by nblum37** [comment]
Pls find the logs in your email box. If you need logs of any other scenarios, pls let me know.

**Comment by blaugold** [comment]
Thanks!

**Comment by nblum37** [comment]
I just read through your cbl-c issue. 
So yes, as you mentioned, I access `document_type` and in very rare cases also `asset_id` in the pull and push filter (same filter algorithm). In very rare cases, the filter causes a Fatal Crash of the App:
<img width="1840" alt="Bildschirmfoto 2021-10-19 um 23 12 07" src="https://user-images.githubusercontent.com/82163574/159131938-de5d63a4-8c55-4d0c-9196-0d32c443c126.png">
But this crash is very rare, not reproducible on purpose and the screenshot is from 19.10.2021. So not sure, if it is somehow connected or relevant.

**Comment by blaugold** [comment]
Those logs were really helpful. Thanks again.

I don't think those crashes are related to this specific issue, and since when the screenshot was captured, a lot has changed, but it would be great if you could open a new issue the next time you come across this kind of crash. Optimally, with the stack trace in text format, so it's easier to symbolicate it.

**Comment by blaugold** [comment]
I published `cbl` `1.0.5` with a fix for this issue.

**Comment by nblum37** [comment]
I included the new fix and everything is working fine. Thank you :)
